squad/core/models.py

```diff
@@ -1114,38 +1108,24 @@ def __init__(self, build, environment=None):
         self.tests_fail = 0
         self.tests_xfail = 0
         self.tests_skip = 0
-        self.failures = OrderedDict()
 
         query_set = build.test_runs
         if environment:
             query_set = query_set.filter(environment=environment)
-        test_runs = query_set.prefetch_related(
-            'environment',
-            'tests',
-            'tests__suite'
-        ).order_by('id')
-
-        tests = OrderedDict()
-        for run in test_runs.all():
-            for test in run.tests.all():
-                tests[(run.environment, test.suite, test.name)] = test
-
-        for context, test in tests.items():
-            if test.status == 'pass':
-                self.tests_pass += 1
-            elif test.status == 'fail':
-                self.tests_fail += 1
-            elif test.status == 'xfail':
-                self.tests_xfail += 1
-            elif test.status == 'skip':
-                self.tests_skip += 1
-            else:
-                raise InvalidTestStatus(test.status)
-            if test.status == 'fail':
-                env = test.test_run.environment.slug
-                if env not in self.failures:
-                    self.failures[env] = []
-                self.failures[env].append(test)
+
+        # 'build' needs to be explicitly passed as one of the wanted fields, otherwise Django
+        # will generate one query per testrun to fetch it's build, even though we're not using it.
+        # We actually only the the id
+        query_set = query_set.only('build').order_by('id')
```

Comment:
in my local tests, adding `only('build')` is what made several queries be executed! e.g.:

```
 build = Build.objects.last()
In [2]: query_set = build.test_runs
In [3]: query_set.all()

<< 1 query >>

In [4]: query_set.order_by('id')

<< 1 query (same as above plus the ORDER BY) >>

Out[4]: SELECT "core_testrun"."id",
       "core_testrun"."build_id",
       "core_testrun"."environment_id",
       "core_testrun"."created_at",
       "core_testrun"."completed",
       "core_testrun"."datetime",
       "core_testrun"."build_url",
       "core_testrun"."job_id",
       "core_testrun"."job_status",
       "core_testrun"."job_url",
       "core_testrun"."resubmit_url",
       "core_testrun"."data_processed",
       "core_testrun"."status_recorded"
  FROM "core_testrun"
 WHERE "core_testrun"."build_id" = 126
 ORDER BY "core_testrun"."id" ASC
 LIMIT 21


Execution time: 0.002020s [Database: default]

<QuerySet [<TestRun: #10593>, <TestRun: #10595>, <TestRun: #10594>, <TestRun: #10596>, <TestRun: #10597>, <TestRun: #10598>, <TestRun: #10599>, <TestRun: #10600>, <TestRun: #10601>, <TestRun: #10602>, <TestRun: #10603>, <TestRun: #10604>, <TestRun: #10605>, <TestRun: #10606>, <TestRun: #10607>, <TestRun: #10608>, <TestRun: #10609>, <TestRun: #10610>, <TestRun: #10612>, <TestRun: #10611>, '...(remaining elements truncated)...']>

In [5]: query_set.only('build').order_by('id')
<< several queries. one like this: >>

Out[5]: SELECT "core_testrun"."id",
       "core_testrun"."build_id"
  FROM "core_testrun"
 WHERE "core_testrun"."build_id" = 126
 ORDER BY "core_testrun"."id" ASC
 LIMIT 21


Execution time: 0.001103s [Database: default]

<< then A LOT like this: >>

SELECT "core_testrun"."id",
       "core_testrun"."job_id"
  FROM "core_testrun"
 WHERE "core_testrun"."id" = 10614


Execution time: 0.000244s [Database: default]

<QuerySet [<TestRun: #10593>, <TestRun: #10595>, <TestRun: #10594>, <TestRun: #10596>, <TestRun: #10597>, <TestRun: #10598>, <TestRun: #10599>, <TestRun: #10600>, <TestRun: #10601>, <TestRun: #10602>, <TestRun: #10603>, <TestRun: #10604>, <TestRun: #10605>, <TestRun: #10606>, <TestRun: #10607>, <TestRun: #10608>, <TestRun: #10609>, <TestRun: #10610>, <TestRun: #10612>, <TestRun: #10611>, '...(remaining elements truncated)...']>
```

Reply:
Fixed it, with `models.TestRun.objects.filter(build=build).only('id')` instead of 'build.test_runs.only('id')'.

Reply:
that still generates 1 extra query for each test run. You want values():

```python
query_set.values('id').order_by('id')
```

the result is then a list of dict's, not TestRun objects.

Reply:
Updated to use values now : )